In [1]:
from Bio import SeqIO
import tqdm

# OLD 2024 MODEL (YF_IDs)

In [2]:
with open('./tgd2024/Manual_check-total-gene.gff3_Right_UTR.gff3_cds.fasta', 'r') as f:
    cds_records = list(SeqIO.parse(f, 'fasta'))

cds = [r.id for r in cds_records]
cds[0], len(cds), len(set(cds))

('YF00013476.t1', 25987, 25987)

In [3]:
with open('./tgd2024/Manual_check-total-gene.gff3_Right_UTR.gff3_exon.fasta', 'r') as f:
    exon_records = list(SeqIO.parse(f, 'fasta'))

exon = [r.id for r in exon_records]
exon[0]
exon[0], len(exon), len(set(exon))

('YF00013476.t1', 26749, 26749)

### THE DIFFERENCE BETWEEN THE EXON AND CDS FILES

In [4]:
len(set.difference(set(exon), set(cds))), len(set.difference(set(cds), set(exon)))

(1104, 342)

1104 more genes in the exon file.
342 more genes in the cds file.

In [5]:
set.difference(set(exon), set(cds))

{'YF00037736.t1',
 'YF00037424.t1',
 'YF00034783.t1',
 'YF00036815.t1',
 'YF00035494.t1',
 'YF00034603.t1',
 'YF00035511.t1',
 'YF00034785.t1',
 'YF00036730.t1',
 'YF00035126.t1',
 'YF00038055.t1',
 'YF00036110.t1',
 'YF00036932.t1',
 'YF00036415.t1',
 'YF00037115.t1',
 'YF00036882.t1',
 'YF00035152.t1',
 'YF00035287.t1',
 'YF00037094.t1',
 'YF00038129.t1',
 'YF00034676.t1',
 'YF00036220.t1',
 'YF00035967.t1',
 'YF00036068.t1',
 'YF00036146.t1',
 'YF00036696.t1',
 'YF00037779.t1',
 'YF00036810.t1',
 'YF00037507.t1',
 'YF00035954.t1',
 'YF00037412.t1',
 'YF00036368.t1',
 'YF00037469.t1',
 'YF00038183.t1',
 'YF00038317.t1',
 'YF00037183.t1',
 'YF00036848.t1',
 'YF00036627.t1',
 'YF00037757.t1',
 'YF00038171.t1',
 'YF00035709.t1',
 'YF00036974.t1',
 'YF00034533.t1',
 'YF00036606.t1',
 'YF00037927.t1',
 'YF00036144.t1',
 'YF00034673.t1',
 'YF00036188.t1',
 'YF00037030.t1',
 'YF00036751.t1',
 'YF00036605.t1',
 'YF00035649.t1',
 'YF00035712.t1',
 'YF00035917.t1',
 'YF00037688.t1',
 'YF000355

In [6]:
set.difference(set(cds), set(exon))

{'YF00034729.t1',
 'YF00034884.t1',
 'YF00035084.t1',
 'YF00035086.t1',
 'YF00035093.t1',
 'YF00035114.t1',
 'YF00035132.t1',
 'YF00035136.t1',
 'YF00035163.t1',
 'YF00035177.t1',
 'YF00035203.t1',
 'YF00035243.t1',
 'YF00035247.t1',
 'YF00035249.t1',
 'YF00035279.t1',
 'YF00035288.t1',
 'YF00035307.t1',
 'YF00035318.t1',
 'YF00035331.t1',
 'YF00035336.t1',
 'YF00035349.t1',
 'YF00035368.t1',
 'YF00035376.t1',
 'YF00035390.t1',
 'YF00035399.t1',
 'YF00035407.t1',
 'YF00035409.t1',
 'YF00035417.t1',
 'YF00035420.t1',
 'YF00035431.t1',
 'YF00035456.t1',
 'YF00035461.t1',
 'YF00035464.t1',
 'YF00035468.t1',
 'YF00035488.t1',
 'YF00035497.t1',
 'YF00035503.t1',
 'YF00035512.t1',
 'YF00035527.t1',
 'YF00035536.t1',
 'YF00035567.t1',
 'YF00035572.t1',
 'YF00035582.t1',
 'YF00035598.t1',
 'YF00035607.t1',
 'YF00035613.t1',
 'YF00035616.t1',
 'YF00035623.t1',
 'YF00035629.t1',
 'YF00035643.t1',
 'YF00035647.t1',
 'YF00035650.t1',
 'YF00035655.t1',
 'YF00035665.t1',
 'YF00035667.t1',
 'YF000356

In both cases, the differing genes are unique YF_IDs.

### THE INTERSECTION BETWEEN THE PEPTIDE AND CDS FILE

In [7]:
with open('./tgd2024/Manual_check-total-gene.gff3_Right_UTR.gff3_pep.fasta', 'r') as f:
    pep_records = list(SeqIO.parse(f, 'fasta'))

pep = [r.id for r in pep_records]
pep[0]
pep[0], len(pep), len(set(pep))

('YF00013476.t1', 25987, 25987)

In [8]:
len(set.intersection(set(pep), set(cds))), len(set(pep)), len(set(cds))

(25987, 25987, 25987)

The genes in the two files are identical.

# NEW 2024 MODEL (TTHERM_IDs) (MAY 15)

In [9]:
with open('./tgd2024_may15/Tetrahymena_Genome_annotation_V2024_CDS.fasta', 'r') as f:
    cds_may15_records = list(SeqIO.parse(f, 'fasta'))

cds_may15 = [r.id for r in cds_may15_records]
cds_may15[0]
cds_may15[0], len(cds_may15), len(set(cds_may15))

('TTHERM_00161861', 26686, 26686)

In [10]:
with open('./tgd2024_may15/Tetrahymena_Genome_annotation_V2024_mRNA.fasta', 'r') as f:
    mrna_may15_records = list(SeqIO.parse(f, 'fasta'))

mrna_may15 = [r.id for r in mrna_may15_records]
mrna_may15[0], len(mrna_may15), len(set(mrna_may15))

('TTHERM_00161861', 53372, 53372)

### THE DIFFERENCE BETWEEN THE EXON AND CDS FILES

In [11]:
len(set.difference(set(mrna_may15), set(cds_may15))), len(set(mrna_may15)), len(set(cds_may15))

(26686, 53372, 26686)

The mrna file contains twice as many genes.

In [12]:
set.difference(set(mrna_may15), set(cds_may15))

{'g16360.t1',
 'g550.t1',
 'g25973.t1',
 'g29305.t1',
 'g24943.t1',
 'g5371.t1',
 'g23966.t1',
 'g21538.t1',
 'g10880.t1',
 'g16023.t1',
 'g10303.t1',
 'g29366.t1',
 'g15549.t1',
 'g17260.t1',
 'g7103.t1',
 'g18547.t1',
 'g21735.t1',
 'g23815.t1',
 'g20290.t1',
 'g14847.t1',
 'g10671.t1',
 'g10030.t1',
 'g11049.t1',
 'g1548.t1',
 'g11694.t1',
 'g20644.t1',
 'g3827.t1',
 'g5039.t1',
 'g12095.t1',
 'g15568.t1',
 'g828.t1',
 'g2089.t1',
 'g25333.t1',
 'g1664.t1',
 'g12126.t1',
 'g3342.t1',
 'g14067.t1',
 'g18800.t1',
 'g14472.t1',
 'g4705.t1',
 'g18652.t1',
 'g26825.t1',
 'g16604.t1',
 'g526.t1',
 'g29233.t1',
 'g21282.t1',
 'g492.t1',
 'g2157.t1',
 'g27265.t1',
 'g19858.t1',
 'g15288.t1',
 'g19867.t1',
 'g3374.t1',
 'g20156.t1',
 'g7251.t1',
 'g17567.t1',
 'g5213.t1',
 'g28112.t1',
 'g26521.t1',
 'g28606.t1',
 'g6405.t1',
 'g2007.t1',
 'g15848.t1',
 'g18917.t1',
 'g27805.t1',
 'g24815.t1',
 'g16525.t1',
 'g4731.t1',
 'g6637.t1',
 'g1574.t1',
 'g594.t1',
 'g1892.t1',
 'g19951.t1',
 'g1391

All of the duplicated genes are gids. I wonder if all of the sequences are duplicated.

In [13]:
mrna_may15_g_seqs = {r.id: str(r.seq) for r in mrna_may15_records if 'g' in r.id}
mrna_may15_ttherm_seqs = {r.id: str(r.seq) for r in mrna_may15_records if 'TTHERM_' in r.id}

In [14]:
len(mrna_may15_g_seqs), list(mrna_may15_g_seqs.keys())[0], len(mrna_may15_ttherm_seqs), list(mrna_may15_ttherm_seqs.keys())[0]

(26686, 'g18410.t1', 26686, 'TTHERM_00161861')

In [15]:
sorted(list(mrna_may15_g_seqs.values()))[0]

'AAAAAAAAAAAAAAAAATTAAGCAAGGCCTACTTTATTTTTAAAAATTTTGAAAAACTTATTTTCCATTTTATTATAAACAGAGTATGAAAACAAATTAGCTTTGTCAAATTGATATTCCTAAAAATTAATATTCAATTAATGCGCCATCAAACTAACACTATAATTATAAATATCTAAAGGAAAAAATTCTACCTGGCCTTACTCCTATAAACTCTCCTGATTTATCACCTTAGATATCATTAAATTAGATAGAAGATCCAGCATTTGCAAATATTAAAATGCTAAAAATTGATTAAGAAAGTAAGAATAATTTTTCTATGAACTAATTTTCTCAATTAAGATATGGCAGCAGAGTTATACATAAAGAACAACATAAGGCTGAGTTTGATAATCTAAATAATTGTGAGGTTCTGCTAACTATTATAAAAAATTTATACAAAATATTCAACACCCTTAAAATAATAGGCAGAAAACGATGAGAAAGAGTAATAAACAGGTATTGTACATGTTTCTAATTACCATAAAAAATAAACTGAAGCCAAAAAATCAGAAGACCTTGCTTCAGAGCTATAAGGATTTGAATAAGAAGTGTATGGTACAAGCAACAAATCTACTGAATCAAAATTTAGTGTTTTACAAAAATTTAGTACAACCAATAATAATAAATTAGAATTTCAAGACAATTCTTTAGAGTTTTACAATATGTAAATTATCTTATCTTTTTTTTACACATAAAGTTTATGGAAATTAAATTAAGTTGAATTTTATTAAACTATGACAAATAATAATTAATCATAATTTACCAATAATTTAATAGATTTAGGATTTAATTAATAGCTATGCTATGATAATGTTTTATTGTGATTTAAAAAATAATAAATATTTATCTAAAAGCTGGGTTACTTTTAAATAATTTGTTGAAAATTTAATATATTTTTTTTAAAAGTTGATTTTATTTTTATTTTGGAGTGATAAAAATAAATTTGAAAAAACATTA

In [16]:
sorted(list(mrna_may15_ttherm_seqs.values()))[0]

'AAAAAAAAAAAAAAAAATTAAGCAAGGCCTACTTTATTTTTAAAAATTTTGAAAAACTTATTTTCCATTTTATTATAAACAGAGTATGAAAACAAATTAGCTTTGTCAAATTGATATTCCTAAAAATTAATATTCAATTAATGCGCCATCAAACTAACACTATAATTATAAATATCTAAAGGAAAAAATTCTACCTGGCCTTACTCCTATAAACTCTCCTGATTTATCACCTTAGATATCATTAAATTAGATAGAAGATCCAGCATTTGCAAATATTAAAATGCTAAAAATTGATTAAGAAAGTAAGAATAATTTTTCTATGAACTAATTTTCTCAATTAAGATATGGCAGCAGAGTTATACATAAAGAACAACATAAGGCTGAGTTTGATAATCTAAATAATTGTGAGGCAGAAAACGATGAGAAAGAGTAATAAACAGGTATTGTACATGTTTCTAATTACCATAAAAAATAAACTGAAGCCAAAAAATCAGAAGACCTTGCTTCAGAGCTATAAGGATTTGAATAAGAAGTGTATGGTACAAGCAACAAATCTACTGAATCAAAATTTAGTGTTTTACAAAAATTTAGTACAACCAATAATAATAAATTAGAATTTCAAGACAATTCTTTAGAGTTTTACAATATAAATGAAACCTAAAAGAAAAAAGTTATGGAATTAAAAAGTATTATTGCTTAAAAAATAGAAACAAACACTTTATAACAGAAAATATTCAAAAAATTTGATAATTTTACTATTTTAAAATTTTTGAATGCAAGAGATGGATCGATAAAGGATGGTTGCTAGATGTTTATAGATTTTTTATAATGGAGAATAGATAATTAAGTTGAAAATATTAACGAATTTTAGTTTTAGGAGTACGATTAAGTTTAGAATGTTTACCCTCACGGATTCCATGGGTATGATAATGAAGGAAGACCAATTTGGATTGAAAATTTAGGAAAGTTAAAATTAAAAGAGTTGATGAAAATAACGAA

In [17]:
len(sorted(list(mrna_may15_g_seqs.values()))[0]), len(sorted(list(mrna_may15_ttherm_seqs.values()))[0])

(2128, 1580)

In [18]:
sorted(list(mrna_may15_g_seqs.values()))[0] == sorted(list(mrna_may15_ttherm_seqs.values()))[0]

False

In [19]:
not_equal = 0

for idx, id in tqdm.tqdm(enumerate(sorted(list(mrna_may15_g_seqs.values())))):
    if id != sorted(list(mrna_may15_ttherm_seqs.values()))[idx]:
        not_equal += 1

not_equal

26686it [03:42, 120.08it/s]


26666

The sequences are not duplicated.

### THE INTERSECTION BETWEEN THE PEPTIDE AND CDS FILE

In [20]:
with open('./tgd2024_may15/Tetrahymena_Genome_annotation_V2024_Protein.fasta', 'r') as f:
    pep_may15_records = list(SeqIO.parse(f, 'fasta'))

pep_may15 = [r.id for r in pep_may15_records]
pep_may15[0], len(pep_may15), len(set(pep_may15))

('TTHERM_00161861', 26686, 26686)

In [21]:
len(set.intersection(set(pep_may15), set(cds_may15))), len(set(pep_may15)), len(set(cds_may15))

(26686, 26686, 26686)

The genes in the two files are identical.